In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

df = pd.read_csv("../data/raw/train.csv")
df["LogSalePrice"] = np.log(df["SalePrice"])

## Stage 4 — Linear Regression Analysis

In this stage, linear regression is used to model the relationship between house prices and selected numerical features.

The goal is not prediction, but understanding:
- how variables are related to house prices
- how to interpret regression coefficients
- how regression connects to correlation results from the previous stage

### Simple Linear Regression: Price and Overall Quality

Research question:  
How does the overall quality of a house relate to its sale price?

In [3]:
!pip install statsmodels

   ---------------------------------------- 0.0/9.6 MB ? eta -:--:--
   ------ --------------------------------- 1.6/9.6 MB 9.3 MB/s eta 0:00:01
   -------------- ------------------------- 3.4/9.6 MB 9.1 MB/s eta 0:00:01
   ------------------- -------------------- 4.7/9.6 MB 7.7 MB/s eta 0:00:01
   ------------------------ --------------- 5.8/9.6 MB 7.3 MB/s eta 0:00:01
   ----------------------------- ---------- 7.1/9.6 MB 7.0 MB/s eta 0:00:01
   ----------------------------------- ---- 8.4/9.6 MB 6.8 MB/s eta 0:00:01
   ---------------------------------------  9.4/9.6 MB 6.6 MB/s eta 0:00:01
   ---------------------------------------- 9.6/9.6 MB 6.5 MB/s  0:00:01

   -------------------- ------------------- 1/2 [statsmodels]
   -------------------- ------------------- 1/2 [statsmodels]
   -------------------- ------------------- 1/2 [statsmodels]
   -------------------- ------------------- 1/2 [statsmodels]
   -------------------- ------------------- 1/2 [statsmodels]
   ------------

In [ ]:
import statsmodels.api as sm

X = df["OverallQual"]
y = df["LogSalePrice"]

X = sm.add_constant(X)  

model = sm.OLS(y, X).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:           LogSalePrice   R-squared:                       0.668
Model:                            OLS   Adj. R-squared:                  0.668
Method:                 Least Squares   F-statistic:                     2931.
Date:                Mon, 09 Feb 2026   Prob (F-statistic):               0.00
Time:                        19:06:25   Log-Likelihood:                 73.088
No. Observations:                1460   AIC:                            -142.2
Df Residuals:                    1458   BIC:                            -131.6
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const          10.5844      0.027    388.183      0.000      10.531      10.638
OverallQual     0.2360      0.004     54.137      0.000       0.227       0.245
==============================================================================
Omnibus:                       87.973   Durbin-Watson:                   1.972
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              165.054
Skew:                          -0.425   Prob(JB):                     1.44e-36
Kurtosis:                       4.410   Cond. No.                         29.0
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Interpretation of Regression Results

A simple linear regression was fitted using `OverallQual` to explain `LogSalePrice`.

The regression coefficient for `OverallQual` is positive and statistically significant.
This means that houses with higher overall quality tend to have higher sale prices.

The coefficient value (about 0.24) indicates that when the overall quality increases by one unit,
the log-transformed sale price increases by about 0.24.
Because the target variable is in logarithmic form, this can be interpreted as an approximate percentage increase in price.

The R-squared value is about 0.67.
This means that around 67% of the variation in house prices is explained by overall quality alone.

This result shows a strong relationship between house quality and sale price.
However, this model includes only one variable and does not account for other factors such as size, location, or age of the house.

### Multiple Linear Regression: Overall Quality and Living Area

After fitting a simple linear regression model, the analysis is extended to a multiple linear regression model.

In this model, two explanatory variables are used:
- `OverallQual`, which represents the overall quality of the house
- `GrLivArea`, which represents the above-ground living area

The goal of this step is to examine how these variables jointly relate to house prices.
Including more than one variable allows us to see the effect of each variable while holding the other one constant.

The dependent variable remains `LogSalePrice`, which helps reduce skewness and makes the interpretation more stable.

In [5]:
import statsmodels.api as sm

X = df[["OverallQual", "GrLivArea"]]
y = df["LogSalePrice"]

X = sm.add_constant(X)

model_multi = sm.OLS(y, X).fit()
model_multi.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:           LogSalePrice   R-squared:                       0.740
Model:                            OLS   Adj. R-squared:                  0.740
Method:                 Least Squares   F-statistic:                     2073.
Date:                Mon, 09 Feb 2026   Prob (F-statistic):               0.00
Time:                        19:11:46   Log-Likelihood:                 251.92
No. Observations:                1460   AIC:                            -497.8
Df Residuals:                    1457   BIC:                            -482.0
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const          10.5486      0.024    435.947      0.000      10.501      10.596
OverallQual     0.1789      0.005     37.328      0.000       0.169       0.188
GrLivArea       0.0003   1.26e-05     20.111      0.000       0.000       0.000
==============================================================================
Omnibus:                      480.162   Durbin-Watson:                   1.981
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3360.538
Skew:                          -1.348   Prob(JB):                         0.00
Kurtosis:                       9.926   Cond. No.                     7.35e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 7.35e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### Interpretation of Multiple Regression Results

A multiple linear regression model was fitted using `OverallQual` and `GrLivArea` to explain `LogSalePrice`.

Both variables have positive and statistically significant coefficients.
This means that house quality and living area are both related to higher house prices.

The coefficient for `OverallQual` is smaller than in the simple regression model.
This shows that part of the effect of quality is shared with house size.
However, even after accounting for living area, overall quality remains an important factor in explaining house prices.

The coefficient for `GrLivArea` is also positive.
This indicates that, for houses with the same quality, larger living area is associated with higher prices.

The R-squared value is about 0.74.
This means that approximately 74% of the variation in house prices is explained by this model.
This is an improvement compared to the simple regression model.

Overall, the multiple regression model provides a better explanation of house prices than a single-variable model.